gobblin-data-management/src/main/java/gobblin/data/management/retention/version/HiveDatasetVersionCleaner.java

```diff
@@ -0,0 +1,250 @@
+import com.typesafe.config.Config;
+
+import gobblin.data.management.retention.dataset.CleanableDataset;
+import gobblin.data.management.retention.dataset.CleanableHiveDataset;
+import gobblin.data.management.version.DatasetVersion;
+import gobblin.data.management.version.HiveDatasetVersion;
+import gobblin.util.AutoReturnableObject;
+import gobblin.util.ConfigUtils;
+
+
+/**
+ * An abstraction for cleaning a {@link HiveDatasetVersionCleaner} of a {@link CleanableHiveDataset}.
+ */
+@Slf4j
+public class HiveDatasetVersionCleaner extends VersionCleaner {
+
+  public static final String REPLACED_HIVE_DB_NAME_TEMPLATE_KEY = "replacedHiveDbNameTemplate";
+  public static final String REPLACED_HIVE_TABLE_NAME_TEMPLATE_KEY = "replacedHiveTableNameTemplate";
+  public static final String REPLACEMENT_HIVE_DB_NAME_TEMPLATE_KEY = "replacementHiveDbNameTemplate";
+  public static final String REPLACEMENT_HIVE_TABLE_NAME_TEMPLATE_KEY = "replacementHiveTableNameTemplate";
+  public static final String SHOULD_REPLACE_PARTITION_KEY = "shouldReplacePartition";
+
+  public static final String DB_TOKEN = "$DB";
+  public static final String TABLE_TOKEN = "$TABLE";
+
+  private final CleanableHiveDataset cleanableHiveDataset;
+  private final HiveDatasetVersion hiveDatasetVersion;
+
+  private final String replacementDbName;
+  private final String replacementTableName;
+
+  public HiveDatasetVersionCleaner(DatasetVersion datasetVersion, CleanableDataset cleanableDataset) {
+    super(datasetVersion, cleanableDataset);
+
+    Preconditions.checkArgument(cleanableDataset instanceof CleanableHiveDataset, String.format("%s only supports %s, "
+        + "found: %s", this.getClass(), CleanableHiveDataset.class, cleanableDataset.getClass()));
+    Preconditions.checkArgument(datasetVersion instanceof HiveDatasetVersion, String.format("%s only supports %s, "
+        + "found: %s", this.getClass(), HiveDatasetVersionCleaner.class, datasetVersion.getClass()));
+
+    this.cleanableHiveDataset = (CleanableHiveDataset) cleanableDataset;
+    this.hiveDatasetVersion = (HiveDatasetVersion) datasetVersion;
+
+    // For post cleanup activity:
+    // Determine source db / table name from which partition has to be replaced-in for the target partition being deleted.
+    // This is done as follows:
+    // 1. The db and table from which partition is being deleted is assumed to be in the following format:
+    //    {prefix}{name_element}{postfix}
+    // 2. The template (format) as in #1 comes in from config and using it the {name_element} is extracted.
+    //    If no template is specified, the db / table name is assumed as the {name_element} in entirety.
+    // 3. Another template is used to determine the db / table from which replacement partition has to come in.
+    //    This template can be of the format {newPrefix}{name_element}{postFix}
+    String extractedDbName = extractNameFromTemplate(cleanableHiveDataset.getDatasetConfig(),
```

Comment:
As discussed offline, made real and logical db and table names citizen of HiveDataset. Also, added auto-resolution of all config properties with real and logical db and table names. 

As well as, added integration tests.